File: 2130. Maximum Twin Sum of a Linked List/PairSum.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    # O(n) time, 
    # O(1) space,
    # Approach: two pointers, 
    def pairSum(self, head: Optional[ListNode]) -> int:
        slow, fast = head, head
        
        # find the mid
        while fast:
            fast = fast.next.next
            slow = slow.next
            
        nxt = None
        curr = slow.next
        prev = slow
        prev.next = None
        
        while curr:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt
            
        max_twin_sum = float('-inf')
        curr = head
        while prev:
            curr_twin_sum = prev.val + curr.val
            max_twin_sum = max(max_twin_sum, curr_twin_sum)
            prev = prev.next
            curr = curr.next
            
        return max_twin_sum

File: 2130. Maximum Twin Sum of a Linked List/PairSum.java
public class PairSum {

    public class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next;
        }
    }

    public int pairSum(ListNode head) {
        ListNode fast = head;
        ListNode slow = head;

        // find the middle of the linkedlist
        while (fast != null && fast.next != null){
            fast = fast.next.next;
            slow = slow.next;
        }

        // reverse from end to the middle of the linkedlist
        ListNode prev = null;
        while (slow != null){
            ListNode temp = slow.next;
            slow.next = prev;
            prev = slow;
            slow = temp;
        }

        // now calculate the pairsum of each twin nodes
        int maxPairSum = Integer.MIN_VALUE;
        while (prev != null){
            maxPairSum = Math.max(maxPairSum, head.val + prev.val);
            head = head.next;
            prev = prev.next;
        }

        return maxPairSum;
    }
    
}
